packages/adblocker-extended-selectors/src/eval.ts

```diff
@@ -165,99 +171,48 @@ export function matches(element: Element, selector: AST): boolean {
-  return null;
-}
-
-function handleCompoundSelector(element: Element, compound: AST[]): Element[] {
-  if (compound.length === 0) {
-    return [];
-  }
-
-  const upwardIndex = compound.findIndex((s) => s.type === 'pseudo-class' && s.name === 'upward');
-
-  if (upwardIndex === -1) {
-    const [firstSelector, ...restSelectors] = compound;
-    return querySelectorAll(element, firstSelector).filter((e) =>
-      restSelectors.every((s) => matches(e, s)),
-    );
-  }
-  const before = compound.slice(0, upwardIndex);
-  const upward = compound[upwardIndex];
-  const after = compound.slice(upwardIndex + 1);
-
-  const candidates =
-    before.length > 0
-      ? querySelectorAll(element, { type: 'compound', compound: before })
-      : [element];
-
-  const ancestors = new Set<Element>();
-
-  for (const candidate of candidates) {
-    if (upward.type !== 'pseudo-class' || upward.name !== 'upward') {
-      continue;
-    }
-    const { argument } = upward;
-    if (argument === undefined) {
-      continue;
-    }
-
-    const distance = parseInt(argument, 10);
-    const ancestor = !Number.isNaN(distance)
-      ? findAncestorByDistance(candidate, distance)
-      : findAncestorBySelector(candidate, argument);
-
-    if (ancestor === null) {
-      continue;
-    }
-    ancestors.add(ancestor);
-  }
-
-  if (ancestors.size === 0) {
-    return [];
-  }
-
-  if (after.length > 0) {
-    if (after[0].type === 'pseudo-class' && after[0].name === 'upward') {
-      return Array.from(ancestors).flatMap((a) =>
-        querySelectorAll(a, { type: 'compound', compound: after }),
-      );
+/**
+ * Try transposing with a selector from the subjective element.
+ * @param element The subjective element
+ * @param selector A selector
+ * @returns An array with and without singular element; we may support transposing to multiple targets in the future
+ * but currently it's for the convenience to match return type.
+ */
+function transpose(element: Element, selector: AST): [] | [Element] {
+  if (selector.type === 'pseudo-class') {
+    if (selector.name === 'upward') {
+      if (selector.argument === undefined) {
+        return [];
+      }
+      const literalOrNumeric = stripsWrappingQuotes(selector.argument);
+      let number = parseInt(literalOrNumeric, 10);
+      let parentElement: Element | null = element;
+      if (!Number.isInteger(number)) {
+        while ((parentElement = parentElement.parentElement) !== null) {
+          if (parentElement.matches(literalOrNumeric)) {
+            return [parentElement];
+          }
+        }
+      } else {
+        if (number <= 0 || number >= 256) {
+          return [];
+        }
+        while ((parentElement = parentElement.parentElement) !== null) {
+          if (--number === 0) {
+            return [parentElement];
+          }
+        }
+      }
     }
-    return Array.from(ancestors).filter((a) => after.every((s) => matches(a, s)));
   }
 
-  return Array.from(ancestors);
+  return [];
```

Comment:
(this would be where we could return "null" or something else that signals that "transpose" did not detect any operator that it can handle)